base/daemon/SentryWriter.cpp

```diff
@@ -0,0 +1,244 @@
+#    include "Common/config_version.h"
+#    include <Common/config.h>
+#endif
+
+#if USE_SENTRY
+#    include <sentry.h> // Y_IGNORE
+#    include <stdio.h>
+#endif
+
+
+#if USE_SENTRY
+namespace
+{
+
+bool initialized = false;
+bool anonymize = false;
+
+void setExtras()
+{
+
+    if (!anonymize)
+    {
+        sentry_set_extra("server_name", sentry_value_new_string(getFQDNOrHostName().c_str()));
+    }
+    sentry_set_tag("version", VERSION_STRING_SHORT);
+    sentry_set_extra("version_githash", sentry_value_new_string(VERSION_GITHASH));
+    sentry_set_extra("version_describe", sentry_value_new_string(VERSION_DESCRIBE));
+    sentry_set_extra("version_integer", sentry_value_new_int32(VERSION_INTEGER));
+    sentry_set_extra("version_revision", sentry_value_new_int32(VERSION_REVISION));
+    sentry_set_extra("version_major", sentry_value_new_int32(VERSION_MAJOR));
+    sentry_set_extra("version_minor", sentry_value_new_int32(VERSION_MINOR));
+    sentry_set_extra("version_patch", sentry_value_new_int32(VERSION_PATCH));
+}
+
+void sentry_logger(sentry_level_t level, const char * message, va_list args)
+{
+    auto * logger = &Poco::Logger::get("SentryWriter");
+    size_t size = 1024;
+    char buffer[size];
+#ifdef __clang__
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wformat-nonliteral"
+#endif
+    if (vsnprintf(buffer, size, message, args) >= 0)
+    {
+#ifdef __clang__
+#pragma clang diagnostic pop
+#endif
+        switch (level)
+        {
+            case SENTRY_LEVEL_DEBUG:
+                logger->debug(buffer);
+                break;
+            case SENTRY_LEVEL_INFO:
+                logger->information(buffer);
+                break;
+            case SENTRY_LEVEL_WARNING:
+                logger->warning(buffer);
+                break;
+            case SENTRY_LEVEL_ERROR:
+                logger->error(buffer);
+                break;
+            case SENTRY_LEVEL_FATAL:
+                logger->fatal(buffer);
+                break;
+        }
+    }
+}
+}
+#endif
+
+void SentryWriter::initialize(Poco::Util::LayeredConfiguration & config)
+{
+#if USE_SENTRY
+    bool enabled = false;
+    bool debug = config.getBool("send_crash_reports.debug", false);
+    auto * logger = &Poco::Logger::get("SentryWriter");
+    if (config.getBool("send_crash_reports.enabled", false))
+    {
+        if (debug || (strlen(VERSION_OFFICIAL) > 0))
+        {
+            enabled = true;
+        }
+    }
+    if (enabled)
+    {
+        const std::string & endpoint
+            = config.getString("send_crash_reports.endpoint", "https://6f33034cfe684dd7a3ab9875e57b1c8d@o388870.ingest.sentry.io/5226277");
+        const std::string & temp_folder_path
+            = config.getString("send_crash_reports.tmp_path", config.getString("tmp_path", Poco::Path::temp()) + "sentry/");
+        Poco::File(temp_folder_path).createDirectories();
+
+        sentry_options_t * options = sentry_options_new();
```

Comment:
Some methods below may throw exceptions, need to provide exception safety (use unique_ptr with custom deleter). BTW, where do we destroy the object?

Reply:
If we don't have to destroy the object, it should be stated in comments.